keps/sig-node/1948-deallocate-dp-api/README.md

```diff
@@ -0,0 +1,846 @@
+be needed.
+
+This KEP proposes adding two extra API calls, maintaining the same logical
+reasoning of the previous two. Those are:
+
+- `Deallocate`: (Optional). Which is the opposite of allocate, and is needed to
+inform device plugins that some devices are no longer being used (this used to
+happen silently before)
+- `PostStopContainer`: (Optional). Which allows the device plugins to do device
+cleanup, driver unloading, and any other cleanup actions that may be needed.
+
+Since both additions are optional, existing device plugins should continue
+functioning properly with no needed modifications. Only device plugins that wish
+to utilize the new API calls will need to be modified.
+
+### Risks and Mitigations
+
+Only risk is breaking existing device plugins by introducing non optional
+changes. Can be mitigated by enough test coverage.
+
+## Design Details
+
+- Move `PreStartContainer` in `DeviceManager` to be used as a container
+lifecycle hook. (This change isn't truly required, but useful for compatibility
+and organization with the next steps).
+- Add `PostStopContainer` and `Deallocate` calls in the DevicePlugin API.
+- Add `PostStopContainer` as a container lifecycle hook.
+- Add `Deallocate` calls in container manager, taking care to only do so for
+devices that are no longer in the reuse list.
+- Add and modify test cases for both calls.
+- Test with existing device plugins to ensure the changes are non-breaking.
+- Test with new device plugins utilizing such changes to ensure the changes are
+working.
+
+### Test Plan
+
+- Unit tests will be updated to include the new API calls
+- E2E tests should be added with a sample device plugin for verification.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys, specially about the device reuse
+and possible alternatives.
+- Complete implementation for the new API calls.
+- Tests are in Testgrid and linked in KEP.
+
+#### Beta -> GA Graduation
+
+- More rigorous testing as needed/discussed by developers.
+- Larger scale use/testing by interested users with no reported major bugs.
+
+### Upgrade / Downgrade Strategy
+
+As part of the device plugin API, this will follow the same API versioning
+system. This means that it is up to an application (a device plugin) to choose
+the required API version it wants. As long as the cluster has a recent enough
+(to include the required API version) kubernetes, upgrade or downgrades require
+no cluster modifications at all, and are decided on the application level.
+
+### Version Skew Strategy
+
+As part of the device plugin API, this will follow the same API versioning
+system. This means that it is up to an application (a device plugin) to choose
+the required API version it wants. No version skew issues will arise.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
```

Comment:
As with all new features we will need to introduce a feature gate to block out the code in the `kubelet` that implements this feature. Just ensuring that the feature is "opt-in" is not enough. There must be a way to disable the code-paths that introduce any newly implemented logic in the `kubelet` entirely.

Please see https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2902-cpumanager-distribute-cpus-policy-option/README.md?plain=1#L151 for an example.

Independent of this, you seem to be missing quite a few questions from the new PRR requirements. Were you using an outdated template?

Reply:
Mea culpa. Yes, an outdated template.

Reply:
I added the `DeviceManagerDeallocate` feature gate.